File: src/main/java/io/binac/leetcode/RemoveDuplicatesFromSortedList.java
package io.binac.leetcode;

import io.binac.leetcode.util.ListNode;

/**
 * Given a sorted linked list, delete all duplicates such that each element appear only once.
 * <p>
 * <p>Example 1:
 * <blockquote><pre>
 *     Input: 1->1->2
 *     Output: 1->2
 * </blockquote></pre>
 * Example 2:
 * <blockquote><pre>
 *     Input: 1->1->2->3->3
 *     Output: 1->2->3
 * </blockquote></pre>
 */
public class RemoveDuplicatesFromSortedList {
    public static class Solution1 {
        public ListNode deleteDuplicates(ListNode head) {
            if (head == null)
                return null;
            ListNode tail = head;
            for (ListNode p = head.next; p != null; p = p.next) {
                if (p.val != tail.val) {
                    tail.next = p;
                    tail = p;
                }
            }
            tail.next = null;
            return head;
        }
    }
}

File: python3/leetcodepy/remove_duplicates_from_sorted_list.py
"""
Given the head of a sorted linked list, delete all duplicates such that each element appears only once. Return the linked list sorted as well.



Example 1:
https://assets.leetcode.com/uploads/2021/01/04/list1.jpg

Input: head = [1,1,2]
Output: [1,2]

Example 2:
https://assets.leetcode.com/uploads/2021/01/04/list2.jpg

Input: head = [1,1,2,3,3]
Output: [1,2,3]


Constraints:

The number of nodes in the list is in the range [0, 300].
-100 <= Node.val <= 100
The list is guaranteed to be sorted in ascending order.
"""
from typing import Optional

from .utils import ListNode


class Solution1:
    def deleteDuplicates(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head:
            return head
        tail = head
        p = head.next
        while p:
            if p.val != tail.val:
                tail.next = p
                tail = p
            p = p.next
        tail.next = None
        return head
